src/nvim/eval.c

```diff
@@ -16753,6 +16855,83 @@ static void f_visualmode(typval_T *argvars, typval_T *rettv)
     curbuf->b_visual_mode_eval = NUL;
 }
 
+// watcheradd(dict, key, funcref) function
+static void f_watcheradd(typval_T *argvars, typval_T *rettv)
+{
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  if (argvars[0].v_type != VAR_DICT ||
+      argvars[1].v_type != VAR_STRING ||
```

Comment:
Must use `get_tv_string_chk`. `dict[0]` is a perfectly valid construct, where `0` is implicitly casted to string and this is what will be expected here for `watcheradd(dict, 0, function("Foo"))`.